### ISSUE TITLE
win 32 problems

### ISSUE THREAD

**Issue by ispysoftware** [webrtc-uwp/webrtc-uwp-sdk#181]
I'm trying out the win32 build as our app also runs as a console application and as a windows service. It hangs on 

`PeerConnection = new RTCPeerConnection(config);`

I changed the app type to a windows forms app and it got past that line. I figured it's something to do with:

```
config = new Org.WebRtc.WebRtcLibConfiguration
            {
                Queue = Org.WebRtc.EventQueue.GetDefaultForUi()
            };
```
.. in that console apps don't have a UI. Is there a way to create an event queue for a console/ service application? I tried the GetOrCreateThreadQueueByName method but that didn't work (same issue).

The windows app then crashed with  System.AccessViolationException: 'Attempted to read or write protected memory. This is often an indication that other memory is corrupt.' when i create a video device (I'm using a dummy video device and pushing in frames from code, not touching the built in device code) .

`var videoCapturer = VideoCapturer.Create(vcparams);`

stack trace:
```
 at Wrapper.Org_WebRtc.Api.org_webRtc_VideoCapturer_wrapperObserveEvents(IntPtr handle)
   at Org.WebRtc.VideoCapturer.WrapperObserveEvents()
   at Org.WebRtc.VideoCapturer.org_webRtc_VideoCapturer_AdoptFromC(IntPtr handle)
   at Org.WebRtc.VideoCapturer.Create(VideoCapturerCreationParameters params)
```

and if i skip audio/ video dummy device creation it then throws another access violation exception

```
   at Wrapper.Org_WebRtc.Api.zs_Promise_isResolved(IntPtr handle)
   at Wrapper.Org_WebRtc.Helpers.<>c__DisplayClass165_1.<zs_Promise_AdoptFromC>b__1(Object callbackTarget, String methodName, IntPtr callbackHandle)
   at Wrapper.Org_WebRtc.Helpers.EventManager.Target.FireEvent(String methodName, IntPtr handle)
```

(this is the same code that's working on the UWP app) - is the win32 build supposed to be working?

**Comment by robin-raymond** [comment]
It should be but we've not had a good testing application. If you are able to provide a sample project (zip maybe?) then I would be more than willing to look at the issue.

**Comment by ispysoftware** [comment]
Sorry - I'm just getting back to the win32 platform. It's crashing now just calling setup - It looks like CPPWINRT_VERSION is defined - not sure why it's defined in the win32 project..

because it's defined it's running:

#ifdef CPPWINRT_VERSION

  {
    // Setup if WinUWP CppWinRT is defined...
    auto nativeCppWinrt = UseEventQueue::toNative_winrtCoreDispatcher(queue);
    if ((nativeCppWinrt) && (!didSetupZsLib_.test_and_set())) {
      UseHelper::setup(nativeCppWinrt);
    }
  }

#endif //CPPWINRT_VERSION


here's the call stack:

`>	libOrg.WebRtc.WrapperC.dll!winrt::impl::consume_Windows_UI_Core_ICoreWindow<winrt::Windows::UI::Core::ICoreWindow>::Dispatcher() Line 636	C++
 	libOrg.WebRtc.WrapperC.dll!zsLib::internal::MessageQueueThreadUsingCurrentGUIMessageQueueForCppWinrt::singleton() Line 120	C++
 	libOrg.WebRtc.WrapperC.dll!zsLib::internal::MessageQueueThread::singletonUsingCurrentGUIThreadsMessageQueue() Line 177	C++
 	libOrg.WebRtc.WrapperC.dll!zsLib::IMessageQueueThread::singletonUsingCurrentGUIThreadsMessageQueue() Line 273	C++
 	libOrg.WebRtc.WrapperC.dll!wrapper::impl::org::webRtc::WebRtcLib::actual_delegateQueue() Line 344	C++
 	libOrg.WebRtc.WrapperC.dll!wrapper::impl::org::webRtc::WebRtcLib::actual_setup(std::shared_ptr<wrapper::org::webRtc::WebRtcLibConfiguration> configuration) Line 207	C++
 	libOrg.WebRtc.WrapperC.dll!wrapper::impl::org::webRtc::WebRtcLib::actual_setup() Line 162	C++
 	libOrg.WebRtc.WrapperC.dll!wrapper::org::webRtc::WebRtcLib::setup() Line 95	C++
 	libOrg.WebRtc.WrapperC.dll!org_webRtc_WebRtcLib_setup() Line 24	C++
 	[External Code]	
 	Org.WebRtc.dll!Org.WebRtc.WebRtcLib.Setup() Line 39	C#
`

**Comment by ispysoftware** [comment]
It looks like that line:
Queue = Org.WebRtc.EventQueue.GetOrCreateThreadQueueByName("rtc")
doesn't cut it - it needs to be some type of event queue with a windows UI messaging dispatcher under the covers - that isn't going to work for console apps. 

The static constructor for the eventqueuemaker in the win32 project is empty (unlike the uwp version). Any ideas here?

**Comment by ispysoftware** [comment]
I got past some of the errors by adding a check into 

D:\Projects\webrtc\webrtc-uwp-sdk\webrtc\xplatform\webrtc\third_party\idl\zsLib\zsLib\cpp\zsLib_MessageQueueThreadUsingCurrentGUIMessageQueueForCppWinrt.cpp

line 119 - added a check to see if window was returned - it was bugging out here as window didn't exist

```
auto window = winrt::Windows::UI::Core::CoreWindow::GetForCurrentThread();
        if (window) {
          auto dispatcher = window.Dispatcher();
          if (dispatcher) {
            setupCoreDispatcher(dispatcher);
          }
        }
```

**Comment by robin-raymond** [comment]
There should be a method to get the GUI thread already on the EventQueue that will work for the GUI thread. The CoreWindow can't work because that requires your GUI to have been build with UWP.

**Comment by ispysoftware** [comment]
Yeah that doesn't work, either in console or forms applications.  It dies at that same point. Adding in that check for window fixed it. It's now bugging out with

Unhandled exception at 0x750B3562 (KernelBase.dll) in AgentService.exe: 0xC0020001: The string binding is invalid (parameters: 0x8007042B).

```
libOrg.WebRtc.WrapperC.dll!wrapper::IWrapperCallbackEvent::fireEvent(std::shared_ptr<wrapper::IWrapperCallbackEvent> event) Line 5833	C++
 	libOrg.WebRtc.WrapperC.dll!wrapper::org_webRtc_CustomAudioDevice_WrapperObserver::onRequestDeviceState(std::shared_ptr<wrapper::org::webRtc::CustomAudioDeviceRequestStateEvent> evt) Line 367	C++
 	libOrg.WebRtc.WrapperC.dll!wrapper::org::webRtc::CustomAudioDevice::onRequestDeviceState(std::shared_ptr<wrapper::org::webRtc::CustomAudioDeviceRequestStateEvent> evt) Line 136	C++
 	libOrg.WebRtc.WrapperC.dll!wrapper::impl::org::webRtc::CustomAudioDevice::StopPlayout::__l2::<lambda>() Line 713	C++
 	libOrg.WebRtc.WrapperC.dll!zsLib::IMessageQueueMessageClosure<void <lambda>(void) >::processMessage() Line 56	C++
 	libOrg.WebRtc.WrapperC.dll!zsLib::internal::MessageQueue::process() Line 103	C++
 	libOrg.WebRtc.WrapperC.dll!zsLib::internal::MessageQueueThreadBasic::operator()() Line 86	C++

```

**Comment by robin-raymond** [comment]
I will need a sample test app to debug it.

**Comment by ispysoftware** [comment]
here you go - you'll need libOrg.WebRtc.WrapperC.dll built - i assume you have that?
[win32test.zip](https://github.com/webrtc-uwp/webrtc-uwp-sdk/files/3697636/win32test.zip)

**Comment by robin-raymond** [comment]
Yes, I can build that. Thanks.

**Comment by ispysoftware** [comment]
also you might need to apply that fix to zslib i posted earlier or it won't get past config

**Comment by ispysoftware** [comment]
btw, i'd added in the set data by pointer stuff

```
  /// <summary>
            /// Sets the video data.
            /// </summary>
            public void SetData8BitWithPtr(ulong buffer, System.UInt64 length)
            {
                Wrapper.Org_WebRtc.OverrideApi.org_webRtc_VideoData_set_data8bitwithptr(this.native_, buffer, length);
            }
```

to
D:\Projects\webrtc\webrtc-uwp-sdk\webrtc\xplatform\webrtc\sdk\windows\wrapper\override\dotnet\net_override_org_webRtc_VideoData.cs

**Comment by robin-raymond** [comment]
I forgot to ask which branch you are working from when you are testing this feature so I can ensure it works properly.

**Comment by ispysoftware** [comment]
Release 75

**Comment by robin-raymond** [comment]
The issue with the queue was due to it being a console app - there's no dispatcher possible without a message pump. So I fixed one issue with the queuing which allows the default queue to work. There's something else wrong though further along in the code though.

**Comment by ispysoftware** [comment]
Yeah, that's what I posted above 3 days ago
https://github.com/webrtc-uwp/webrtc-uwp-sdk/issues/181#issuecomment-538999223

**Comment by robin-raymond** [comment]
The issue I am seeing is there's clearly a compilation difference between the library and the wrapper so some settings are mismatching and causing the issue I'm facing.

**Comment by robin-raymond** [comment]
For configuration x64 / debug:

```
void wrapper::impl::org::webRtc::RTCConfiguration::wrapper_init_org_webRtc_RTCConfiguration() noexcept
{
  assert(_CrtCheckMemory());

  native_ = make_shared<NativeType>();
^^^^^^^^^^^ MEMORY CORRUPTED HERE ^^^^^
  assert(_CrtCheckMemory());
}
```
NativeType is a ::webrtc::PeerConnectionInterface::RTCConfiguration which has a non trivial constructor (because it contains some container objects).

and the ninja that compiles it:

````
defines = -DUSE_AURA=1 -DNO_TCMALLOC -DFULL_SAFE_BROWSING -DSAFE_BROWSING_CSD -DSAFE_BROWSING_DB_LOCAL -DCHROMIUM_BUILD -D_HAS_EXCEPTIONS=0 -D__STD_C -D_CRT_RAND_S -D_CRT_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_DEPRECATE -D_ATL_NO_OPENGL -D_WINDOWS -DCERT_CHAIN_PARA_HAS_EXTRA_FIELDS -DPSAPI_VERSION=2 -DWIN32 -D_SECURE_ATL -D_USING_V110_SDK71_ -DWINAPI_FAMILY=WINAPI_FAMILY_DESKTOP_APP -DWIN32_LEAN_AND_MEAN -DNOMINMAX -D_UNICODE -DUNICODE -DNTDDI_VERSION=NTDDI_WIN10_RS2 -D_WIN32_WINNT=0x0A00 -DWINVER=0x0A00 -D_DEBUG -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DWTF_USE_DYNAMIC_ANNOTATIONS=1 -D_HAS_ITERATOR_DEBUGGING=0 -DWEBRTC_ENABLE_PROTOBUF=1 -DWEBRTC_INCLUDE_INTERNAL_AUDIO_DEVICE -DRTC_ENABLE_VP9 -DHAVE_SCTP -DWEBRTC_LIBRARY_IMPL -DWEBRTC_NON_STATIC_TRACE_EVENT_HANDLERS=0 -DWEBRTC_WIN -DABSL_ALLOCATOR_NOTHROW=1
include_dirs = -I../.. -Igen -I../../third_party/abseil-cpp
cflags = /Gy /FS /bigobj /utf-8 /Zc$:sizedDealloc- /wd4117 /D__DATE__= /D__TIME__= /D__TIMESTAMP__= /W4 /WX /wd4091 /wd4127 /wd4251 /wd4275 /wd4312 /wd4324 /wd4351 /wd4355 /wd4503 /wd4589 /wd4611 /wd4100 /wd4121 /wd4244 /wd4505 /wd4510 /wd4512 /wd4610 /wd4838 /wd4995 /wd4996 /wd4456 /wd4457 /wd4458 /wd4459 /wd4200 /wd4201 /wd4204 /wd4221 /wd4245 /wd4267 /wd4305 /wd4389 /wd4702 /wd4701 /wd4703 /wd4661 /wd4706 /wd4715 /wd4702 /Od /Ob0 /GF /Zi /MTd /wd4702 /wd4267
cflags_cc = /TP /wd4577
label_name = libjingle_peerconnection_api
target_out_dir = obj/api
target_output_name = libjingle_peerconnection_api
````

So there's something different in a define or a compilation setting that is causing this:

The wrapper DLL is compiled with:
````
ORG_WEBRTC_WRAPPER_C_GENERATED_IMPLEMENTATION;WEBRTC_WIN;NOMINMAX;_UNICODE;UNICODE;USE_BUILTIN_SW_CODECS;_HAS_EXCEPTIONS=0;_ITERATOR_DEBUG_LEVEL=0;_DEBUG;ORGWEBRTCC_EXPORTS;_WINDOWS;_USRDLL;%(PreprocessorDefinitions)
````

The wrapper glue is compiled with:
````
WEBRTC_WIN;NOMINMAX;_UNICODE;UNICODE;USE_BUILTIN_SW_CODECS;_HAS_EXCEPTIONS=0;_ITERATOR_DEBUG_LEVEL=0;_DEBUG;_LIB;%(PreprocessorDefinitions)
````

So now it's a matter of figuring out what is important that would cause the header file to be different. HAS_EXCEPTIONS being different would cause it but that appears to be identical across the projects.

**Comment by ispysoftware** [comment]
That's weird 'cause my code is getting past that RTCConfiguration constructor fine - it's crashing  when I call

var videoCapturer = VideoCapturer.Create(vcparams);

on line 282 of my test application

.. i've stepped through the debugger and it's failing on the assert on line 217 of 
D:\Projects\webrtc\webrtc-uwp-sdk\webrtc\xplatform\webrtc\sdk\windows\wrapper\impl_org_webRtc_VideoCapturer.cpp

auto capturer = (dynamic_cast<webrtc::VideoCapturer*>(native_.get()));
ZS_ASSERT(capturer);

**Comment by robin-raymond** [comment]
If you do assert(_CrtCheckMemory()); in that RTCConfiguration method, does it pass for you? We've replicated the issue on several machines.

**Comment by ispysoftware** [comment]
_CrtCheckMemory() is returning 1 on both lines - i'm guessing that's a bad thing.

One thing i noticed is that the set/get data with pointer methods are missing in the videodata  override classes. Not sure if that could be causing problems.

**Comment by ispysoftware** [comment]
removing the video and audio stuff from it it's still crashing before it sets up a connection so it seems you're right that there's deeper issues.  This stuff with ninja and code generation and defines is beyond me - are you guys looking into it?

**Comment by robin-raymond** [comment]
Yes, some progress has been made. A branch with some fixes was created but it's not fully there just yet.

**Comment by ispysoftware** [comment]
Awesome thank you

**Comment by ispysoftware** [comment]
https://github.com/webrtc-uwp/webrtc-uwp-sdk/commit/bf8a26e8215d06198398d06f1bbd6fec13ce6e3b

That's resolved one of the issues with the win32 project (video device creation)
I'm now seeing:

Unable to find an entry point named 'org_webRtc_VideoData_size' in DLL 'libOrg.WebRtc.WrapperC.dll'.

- i'm guessing it should be:

org_webRtc_VideoData_get_size 

I'm now getting a crash 
Exception thrown: read access violation. mtx was 0xDDDDDDF5.

It's coming from a mutex lock in zsLib_Promise in 

bool Promise::isResolved() const noexcept
  {
    AutoRecursiveLock lock(mLock);
    return PromiseState_Resolved == mState;
  }

It's a weird one as if I step through the code in the debugger sometimes it works (about 20% of the time) - so it appears to be a timing issue. It happens just after SuccessAnswer is fired and is coming from the MessageQueueThreadBasic thread (i can't see what it's related to in the debugger, just handles and pointers). Here's the call stack:

 	libOrg.WebRtc.WrapperC.dll!mtx_do_lock(_Mtx_internal_imp_t * mtx, const xtime * target) Line 89	C++
 	libOrg.WebRtc.WrapperC.dll!_Mtx_lock(_Mtx_internal_imp_t * mtx) Line 175	C++
>	libOrg.WebRtc.WrapperC.dll!std::_Mutex_base::lock() Line 47	C++
 	libOrg.WebRtc.WrapperC.dll!std::lock_guard<std::recursive_mutex>::lock_guard<std::recursive_mutex>(std::recursive_mutex & _Mtx) Line 436	C++
 	libOrg.WebRtc.WrapperC.dll!zsLib::Promise::isResolved() Line 646	C++
 	libOrg.WebRtc.WrapperC.dll!zs_Promise_isResolved(unsigned int handle) Line 4848	C++
 	[External Code]	
 	Org.WebRtc.dll!Wrapper.Org_WebRtc.Helpers.zs_Promise_AdoptFromC.AnonymousMethod__1(object callbackTarget, string methodName, System.IntPtr callbackHandle) Line 11516	C#
 	Org.WebRtc.dll!Wrapper.Org_WebRtc.Helpers.EventManager.Target.FireEvent(string methodName, System.IntPtr handle) Line 9705	C#
 	Org.WebRtc.dll!Wrapper.Org_WebRtc.Helpers.EventManager.HandleEvent(System.IntPtr handle, string namespaceName, string className, string methodName, System.IntPtr source) Line 9885	C#
 	Org.WebRtc.dll!Wrapper.Org_WebRtc.Helpers.EventManager.HandleEventCallback(System.IntPtr handle) Line 9898	C#
 	[External Code]	
 	libOrg.WebRtc.WrapperC.dll!wrapper::IWrapperCallbackEvent::fireEvent(std::shared_ptr<wrapper::IWrapperCallbackEvent> event) Line 5833	C++
 	libOrg.WebRtc.WrapperC.dll!wrapper::PromiseCallback::onPromiseSettled(std::shared_ptr<zsLib::Promise> promise) Line 7100	C++
 	libOrg.WebRtc.WrapperC.dll!zsLib::Proxy<zsLib::IPromiseDelegate>::Stub_1_onPromiseSettled::processMessage() Line 510	C++
 	libOrg.WebRtc.WrapperC.dll!zsLib::internal::MessageQueue::process() Line 103	C++
 	libOrg.WebRtc.WrapperC.dll!zsLib::internal::MessageQueueThreadBasic::operator()() Line 86	C++
 	[External Code]	
 	libOrg.WebRtc.WrapperC.dll!thread_start<unsigned int (__stdcall*)(void *),1>(void * const parameter) Line 97	C++

**Comment by ispysoftware** [comment]
If I remove the code on the client that sends ice candidates to my app and I wrap  the onoffer request in a couple of delays then it connects successfully (locally). It looks like some task isn't completing in time or something. My code trickles ICE candidates - I noticed that there's a method

get_canTrickleIceCandidates_NotAvailable

- what does NotAvailable mean in that method name - that I can't trickle ICE or that it can't determine whether or not ICE trickle is available or is it some weird double negative?

.. this gets me over the hump with connection ~60% of the time:
public async Task OnOfferRequest(string sdp)
        {
            var ini = new RTCSessionDescriptionInit {Type = RTCSdpType.Offer, Sdp = sdp};
            var rtcDescription = new RTCSessionDescription(ini);
            RTCAnswerOptions answerOptions = new RTCAnswerOptions();

            await PeerConnection.SetRemoteDescription(rtcDescription);
            var answer = await PeerConnection.CreateAnswer(answerOptions);
            await Task.Delay(3000);
            await PeerConnection.SetLocalDescription(answer);
            await Task.Delay(3000);
            OnSuccessAnswer?.Invoke(answer.Sdp);

        }

**Comment by ispysoftware** [comment]
I found this whilst debugging audio problems (as soon as I try to stream audio it crashes). The override in the UTC version takes a ulong buffer size that is missing in the win32 code

`void wrapper::impl::org::webRtc::AudioData::wrapper_init_org_webRtc_AudioData() noexcept
{
  wrapper_dispose();
}`

**Comment by robin-raymond** [comment]
`get_canTrickleIceCandidates_NotAvailable` -- you can ignore this; it does nothing at all. It's part of the official spec but the google version didn't support it (at least not obviously).

**Comment by ispysoftware** [comment]
The mutex memory crash I mentioned above seems to happen when sending in ICE candidates. Sometimes it works and sometimes it doesn't.  When it doesn't I see in the debug log:

Local: ICE state: Connected
(jsep_transport_controller.cc:318): Not adding candidate because the JsepTransport doesn't exist. Ignore it.
Adding 1,1,candidate:2533265093 1 udp 2122194687 192.168.15.14 64421 typ host generation 0 ufrag OKZa network-id 1
(jsep_transport_controller.cc:318): Not adding candidate because the JsepTransport doesn't exist. Ignore it.
Adding 2,2,candidate:4103309814 1 udp 2122262783 2001:8000:1cc1:a900:a033:d14:9049:2fa4 64422 typ host generation 0 ufrag OKZa network-id 2
(jsep_transport_controller.cc:318): Not adding candidate because the JsepTransport doesn't exist. Ignore it.
Exception thrown: read access violation.
_Mtx_internal_imp_t::_get_cs(...)->**** was 0x1CD01.

**Comment by ispysoftware** [comment]
This is the bug i get every time i run the app if I trickle ICE into the peer connection.

![win32bug](https://user-images.githubusercontent.com/800093/67340989-53ea2f00-f561-11e9-911d-2800452a4fbb.gif)

**Comment by ispysoftware** [comment]
@robin-raymond  poking around in zsLib (which says you're the author of it) - it seems that mLock is sometimes not initialized

![image](https://user-images.githubusercontent.com/800093/67372355-59229a80-f5b0-11e9-933a-2f2c13c5792e.png)

**Comment by ispysoftware** [comment]
I think i've fixed it - at least it hasn't crashed since I added a lock and a status check to zs_Promise_AdoptFromC in net_internal_apis.cs

```
            private static object lock_ = new System.Object();
            public static System.Threading.Tasks.Task zs_Promise_AdoptFromC(zs_Promise_t handle)
            {
                return System.Threading.Tasks.Task.Run(() => {
                    if (System.IntPtr.Zero == handle) return null;

                    var target = new object();
                    var tcs = new System.Threading.Tasks.TaskCompletionSource<object>();
                    var instanceId = Wrapper.Org_WebRtc.Api.zs_Promise_wrapperInstanceId(handle);

                    ObserveEvents("::zs", "Promise", instanceId, target, (object callbackTarget, string methodName, callback_event_t callbackHandle) =>
                    {
                        lock (lock_)
                        {
                            if (tcs.Task.Status != System.Threading.Tasks.TaskStatus.RanToCompletion)
                            {
                                bool resolved = Wrapper.Org_WebRtc.Api.zs_Promise_isResolved(handle);
                                if (resolved)
                                {
                                    tcs.SetResult(null);
                                    return;
                                }
                                var exception = zs_PromiseRejectionReason_FromC(handle);
                                if (null != exception)
                                {
                                    tcs.SetException(exception);
                                }
                                else
                                {
                                    tcs.SetCanceled();
                                }
                            }
                        }
                    });

                    var eventObserverHandle = Wrapper.Org_WebRtc.Api.zs_Promise_wrapperObserveEvents(handle);

                    tcs.Task.Wait();

                    Wrapper.Org_WebRtc.Api.callback_wrapperObserverDestroy(eventObserverHandle);
                    Wrapper.Org_WebRtc.Api.zs_Promise_wrapperDestroy(handle);

                    return tcs.Task.Result;
                });
            }
```

**Comment by robin-raymond** [comment]
So for this issue to be true the event handler must be being called more than once from two different threads. Can you confirm by logging the current thread ID / handle combination outside the lock in the `ObserveEvents` callback and seeing if you get the same handle being called with two different thread IDs?

````
                        System.Threading.Thread thread = System.Threading.Thread.CurrentThread;
                        System.String msg = System.String.Format("===> CALLBACK Thread ID {0} with handle {1}\n", thread.ManagedThreadId, handle);
                        System.Console.WriteLine(msg);
````

The handle represents the memory location of the promise so you should not get the same promise firing an event twice with two different thread IDs (unless the promise memory location gets recycled and reassigned to a new promise object but hopefully that won't happen to confuse the outputs).

**Comment by ispysoftware** [comment]
ok i added that code in and removed my stuff - it worked for a while, usually looks like:

===> zs_Promise_AdoptFromC CALLBACK Thread ID 52 with handle 2963941704848

===> zs_PromiseWith_org_webRtc_RTCSessionDescription_void_AdoptFromC CALLBACK Thread ID 52 with handle 2963950195664

===> zs_Promise_AdoptFromC CALLBACK Thread ID 52 with handle 2963941703920

-- ok

then every so often i get a double callback on session description which causes it to fail:

===> zs_Promise_AdoptFromC CALLBACK Thread ID 52 with handle 2963952636656

===> zs_PromiseWith_org_webRtc_RTCSessionDescription_void_AdoptFromC CALLBACK Thread ID 52 with handle 2963952636944

===> zs_PromiseWith_org_webRtc_RTCSessionDescription_void_AdoptFromC CALLBACK Thread ID 52 with handle 2963952636944

-- crash

**Comment by robin-raymond** [comment]
Ok - weird - so since these are from the same thread I wonder if the lock is necessary? I mean it's not going to hurt but maybe it's not the actual solution. I can see why it would crash if called twice. It never should be and that's bizarre. Do you have some updated testing code I can have to get the double event callback happening?

**Comment by ispysoftware** [comment]
I don't because this is in the guts of our main application now. I added the lock just to make sure it wouldn't fire twice from different threads. You're welcome to remote desktop in and take a look if you want. We're having a new issue now where the webrtc code seems to lock the main thread whilst it's setting up the connection and that lock is interfering with some direct show code we've got running and the thread that runs that is crashing randomly during the rtc song and dance. Is there a way to run the webrtc code on it's own thread and not lock the application?

**Comment by ispysoftware** [comment]
Just got another double/ crash in a different callback
===> zs_Promise_AdoptFromC CALLBACK Thread ID 49 with handle 2110811050352
===> zs_Promise_AdoptFromC CALLBACK Thread ID 49 with handle 2110811050352

**Comment by ispysoftware** [comment]
not sure if it's related but about 10% of the time the connection setup fails. It gets stuck indefinitely on 

`await PeerConnection.SetRemoteDescription(rtcDescription);`

.. looking into the threads it's blocked on zs_Promise_AdoptFromC. I'm guessing it's Waiting on a task that's already completed

**Comment by ispysoftware** [comment]
@robin-raymond  We're seeing a consistent issue with webrtc where when it's connecting or disconnecting it's breaking our direct show grabber - it looks like some part of webrtc is doing some direct show device enumeration or capture graph initialization that is breaking our stuff - there's no reason it should be doing this - do you have any idea why/ where that code would be?

**Comment by ispysoftware** [comment]
note to self: the customvideocapturerfactory needs to be disposed in the OnCreateCustomVideoCapturer callback otherwise it is causing random crashes in the datachannel cleanup code  when the peerconnection object is garbage collected (no idea why).

**Comment by ispysoftware** [comment]
We're also getting access violation exceptions when the peer connection is closing in proxy.cc It looks like the thread is being disposed before the datachannel

void SynchronousMethodCall::Invoke(const rtc::Location& posted_from,
                                   rtc::Thread* t) {
  if (t->IsCurrent()) {
    proxy_->OnMessage(nullptr);
  } else {
    t->Post(posted_from, this, 0); <--- here, 
    e_.Wait(rtc::Event::kForever);
  }
}
![image](https://user-images.githubusercontent.com/800093/69385428-42c84580-0cfa-11ea-836b-f93865789eb0.png)

System.AccessViolationException: 'Attempted to read or write protected memory. This is often an indication that other memory is corrupt.'

call stack:

>	libOrg.WebRtc.WrapperC.dll!webrtc::internal::SynchronousMethodCall::Invoke(const rtc::Location & posted_from, rtc::Thread * t) Line 26	C++
 	libOrg.WebRtc.WrapperC.dll!webrtc::MethodCall0<webrtc::DataChannelInterface,void>::Marshal(const rtc::Location & posted_from, rtc::Thread * t) Line 167	C++
 	libOrg.WebRtc.WrapperC.dll!webrtc::DataChannelProxyWithInternal<webrtc::DataChannelInterface>::Close() Line 321	C++
 	libOrg.WebRtc.WrapperC.dll!wrapper::impl::org::webRtc::RTCDataChannel::wrapper_dispose() Line 88	C++
 	libOrg.WebRtc.WrapperC.dll!org_webRtc_RTCDataChannel_wrapperDispose(unsigned __int64 handle) Line 88	C++
 	[External Code]	
 	Org.WebRtc.dll!Org.WebRtc.RTCDataChannel.Dispose(bool disposing) Line 194	C#
 	Org.WebRtc.dll!Org.WebRtc.RTCDataChannel.~RTCDataChannel() Line 202	C#

**Comment by ispysoftware** [comment]
not sure if anyone is working on this project but incase you are..

it also randomly exits in release mode with code 0xc0000409 which apparently means STATUS_STACK_BUFFER_OVERRUN

Faulting module name: libOrg.WebRtc.WrapperC.dll, version: 0.0.0.0, time stamp: 0x5dd4d6a2
Exception code: 0xc0000409
Fault offset: 0x00000000001ba371

**Comment by ispysoftware** [comment]
also, onClose is never fired on the datachannel

**Comment by ispysoftware** [comment]
i've seen in a few places:
ZS_ASSERT(native_);
  if (!native_) return;

- which is crashing the application because native_ is sometimes null. Shouldn't those lines be swapped over?
